hydra-node/src/Hydra/API/ServerOutput.hs

```diff
@@ -161,15 +163,46 @@ prepareServerOutput OutputCBOR response =
         Fanout -> encodedResponse
     TxValid{Hydra.API.ServerOutput.transaction = tx} -> replacedResponse tx
     TxInvalid{Hydra.API.ServerOutput.transaction = tx} -> replacedResponse tx
-    SnapshotConfirmed{} -> encodedResponse
+    SnapshotConfirmed{Hydra.API.ServerOutput.snapshot} ->
+      replacedSnapshotTxResponse (confirmed snapshot)
     GetUTxOResponse{} -> encodedResponse
     InvalidInput{} -> encodedResponse
     Greetings{} -> encodedResponse
-    PostTxOnChainFailed{} -> encodedResponse
+    PostTxOnChainFailed{postChainTx} ->
+      case postChainTx of
+        CloseTx{confirmedSnapshot} ->
+          case confirmedSnapshot of
+            InitialSnapshot{} -> encodedResponse
+            ConfirmedSnapshot{Hydra.Snapshot.snapshot} ->
+              replacedSnapshotTxInPostTxOnChainFailedResponse (confirmed snapshot)
+        ContestTx{confirmedSnapshot} ->
+          case confirmedSnapshot of
+            InitialSnapshot{} -> encodedResponse
+            ConfirmedSnapshot{Hydra.Snapshot.snapshot} ->
+              replacedSnapshotTxInPostTxOnChainFailedResponse (confirmed snapshot)
+        _other -> encodedResponse
     RolledBack -> encodedResponse
  where
   encodedResponse = encode response
+
   replacedResponse tx =
-    case toJSON response of
-      Object km -> encode $ Object $ KeyMap.insert "transaction" (String . decodeUtf8 . Base16.encode $ serialize' tx) km
-      _other -> encodedResponse
+    encodedResponse &
+      atKey "transaction" .~
+        (Just $ String . decodeUtf8 . Base16.encode $ serialize' tx)
+
+  replacedSnapshotTxResponse txs =
+    let cborTxs = Just . Array $ fromList (String . decodeUtf8 . Base16.encode . serialize' <$> txs)
+    in encodedResponse &
+         key "snapshot" .
+         _Object .
+         at "confirmedTransactions"
+         .~ cborTxs
+
+  replacedSnapshotTxInPostTxOnChainFailedResponse txs =
+    let cborTxs = Just . Array $ fromList (String . decodeUtf8 . Base16.encode . serialize' <$> txs)
+    in encodedResponse &
+          key "postChainTx" .
+          key "confirmedSnapshot" .
+          key "snapshot" .
+          _Object .
+          at "confirmedTransactions" .~ cborTxs
```

Comment:
Could: factor out the transaction serialization and inline these updates

I would move the `Tx -> Text` encoding into a single where clause function and then inline `replacedSnapshotTxInPostTxOnChainFailedResponse` et al to not need to invent such long name.